### ISSUE TITLE
New lint suggestion: Lint against `Some(self)` in `Error::source`

### ISSUE THREAD

**Issue by madsmtm** [rust-lang/rust-clippy#11847]
### What it does

This lint prevents the incorrect usage of returning `Some(self)` in `Error::source`.

### Advantage

- Avoid an infinite loop for consumers of `Error::source` that wants to visit all the error sources.

### Example

```rust
impl Error for MyError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        Some(self)
    }
}
```

Could be written correctly as:

```rust
impl Error for MyError {}
```

**Comment by madsmtm** [comment]
(At least I'm fairly sure this is considered incorrect usage, do correct me if I'm wrong!)
